Code/Tools/FBuild/FBuildCore/Graph/NodeGraph.cpp

```diff
@@ -1050,10 +1056,57 @@ void NodeGraph::BuildRecurse( Node * nodeToBuild, uint32_t cost )
         nodeToBuild->SetState( Node::STATIC_DEPS_READY );
     }
 
-    ASSERT( ( nodeToBuild->GetState() == Node::STATIC_DEPS_READY ) ||
+	ASSERT(( nodeToBuild->GetState() == Node::STATIC_DEPS_READY ) ||
+		   ( nodeToBuild->GetState() == Node::PRE_DYN_DEPS_READY ) ||
+           ( nodeToBuild->GetState() == Node::PRE_DYN_DEPS_DONE ) ||
+		   ( nodeToBuild->GetState() == Node::DYNAMIC_DEPS_DONE ));
+
+
+	// update pre-build dynamic dependencies
+	if (nodeToBuild->GetState() == Node::STATIC_DEPS_READY)
+	{
+		// update pre-build dynamic deps
+		bool forceClean = FBuild::Get().GetOptions().m_ForceCleanBuild;
+		
+		if (nodeToBuild->DeterminePreBuildDynamicDependenciesNeedToBuild(forceClean))
+		{
+			// Put the node into the PRE_DYN_DEPS_READY state before queuing the job so that
+			// it goes through the right build steps
+			nodeToBuild->SetState(Node::PRE_DYN_DEPS_READY);
+
+			// Job will get queued and this node will not be re-evaluated until PRE_DYN_DEPS are DONE
+			nodeToBuild->m_RecursiveCost = cost;
+			JobQueue::Get().AddJobToBatch(nodeToBuild);
+			return;
+		}
+		else
+		{
+			// We are ready to check our pre build dynamic dependencies, since we didn't need to build.
+			nodeToBuild->SetState(Node::PRE_DYN_DEPS_READY);
+		}
+	}
+
+	ASSERT( ( nodeToBuild->GetState() == Node::PRE_DYN_DEPS_READY ) ||
+            ( nodeToBuild->GetState() == Node::PRE_DYN_DEPS_DONE ) ||
+            ( nodeToBuild->GetState() == Node::DYNAMIC_DEPS_DONE ));
+
+	// check pre-build dynamic dependencies
+	if (nodeToBuild->GetState() == Node::PRE_DYN_DEPS_READY)
+	{
+		// all pre build dynamic deps done?
+		bool allDependenciesUpToDate = CheckDependencies(nodeToBuild, nodeToBuild->GetPreBuildDynamicDependencies(), cost);
+		if (allDependenciesUpToDate == false)
+		{
+			return; // not ready or failed
+		}
+
+		nodeToBuild->SetState(Node::PRE_DYN_DEPS_DONE);
+	}
+
+    ASSERT( ( nodeToBuild->GetState() == Node::PRE_DYN_DEPS_DONE ) ||
             ( nodeToBuild->GetState() == Node::DYNAMIC_DEPS_DONE ) );
 
-    if ( nodeToBuild->GetState() != Node::DYNAMIC_DEPS_DONE )
+    if ( nodeToBuild->GetState() == Node::PRE_DYN_DEPS_DONE )
```

Comment:
I changed this from != to == to match the pattern in the rest of this state machine. Logic should be identical.